iostat/iostat.go

```diff
@@ -70,29 +72,78 @@ func (iostat *IOSTAT) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricTy
 	if err != nil {
 		return nil, err
 	}
-	metrics := make([]plugin.MetricType, len(mts))
-	for i, m := range mts {
-
-		if v, ok := data[m.Namespace().String()]; ok {
-			metrics[i] = plugin.MetricType{
-				Namespace_: m.Namespace(),
-				Data_:      v,
-				Timestamp_: time.Now(),
+
+	metrics := []plugin.MetricType{}
+
+	for _, mt := range mts {
+		ns := mt.Namespace()
+		if len(ns) < 5 {
+			return nil, fmt.Errorf("Namespace length is too short (len = %d)", len(ns))
+		}
+
+		if ns[4].Value == "*" {
+			if ns[3].Value == deviceMetric {
+				for k, v := range data {
+					dev, err := extractFromNamespace(k, 4)
+					if err != nil {
+						return nil, err
+					}
+
+					ns[4].Value = dev
+
+					metric := plugin.MetricType{
+						Namespace_: ns,
+						Data_:      v,
+						Timestamp_: time.Now(),
+						Tags_:      map[string]string{"dev": dev}}
+					metrics = append(metrics, metric)
+				}
+				return metrics, nil
+			} else {
+				return nil, fmt.Errorf("Dynamic option * not supported for metric %v", ns)
+			}
+		} else {
+			if v, ok := data[mt.Namespace().String()]; ok {
+				metrics = append(metrics, plugin.MetricType{
+					Namespace_: mt.Namespace(),
+					Data_:      v,
+					Timestamp_: time.Now()})
 			}
+
 		}
 	}
 	return metrics, nil
 }
 
 // GetMetricTypes returns the metric types exposed by iostat
 func (iostat *IOSTAT) GetMetricTypes(_ plugin.ConfigType) ([]plugin.MetricType, error) {
-	keys, _, err := iostat.run([]plugin.MetricType{})
+	namespaces, _, err := iostat.run([]plugin.MetricType{})
 	if err != nil {
 		return nil, err
 	}
-	mts := make([]plugin.MetricType, len(keys))
-	for i, k := range keys {
-		mts[i] = plugin.MetricType{Namespace_: core.NewNamespace(strings.Split(strings.TrimPrefix(k, "/"), "/")...)}
+
+	mts := []plugin.MetricType{}
+	metric := plugin.MetricType{}
+	// List of terminal metric names
+	mList := make(map[string]bool)
+	for _, namespace := range namespaces {
+		ns := core.NewNamespace(strings.Split(strings.TrimPrefix(namespace, "/"), "/")...)
+		// terminal metric name
+		mItem := ns[len(ns)-1]
+		if strings.Contains(namespace, deviceMetric) {
```

Comment:
This is not safe check, if metric name or metric subtype will contain `"device"` this if is perhaps triggered accidentally. You can get rid of this problem by first splitting namespace by `"/"` then checking if length of that split is 3 and first element is string you want to match.